logic/path/src/selectors/legacy.rs

```diff
@@ -22,23 +26,50 @@ impl PartialOrd for WeightedChannelPath {
 }
 
 impl Ord for WeightedChannelPath {
+    /// Favors unexplored paths over fully explored paths even when a better
+    /// alternative exists.
+    ///
+    /// Favors longer paths over shorter paths, longer path
+    /// means more privacy.
+    ///
+    /// If both parts are of the same length, favors paths
+    /// with higher weights.
     fn cmp(&self, other: &Self) -> Ordering {
-        self.1.cmp(&other.1)
+        if other.fully_explored == self.fully_explored {
+            match self.path.len().cmp(&other.path.len()) {
+                Ordering::Equal => self.weight.cmp(&other.weight),
+                Ordering::Greater => Ordering::Greater,
+                Ordering::Less => Ordering::Less,
+            }
+        } else if other.fully_explored {
+            Ordering::Less
+        } else {
+            Ordering::Greater
+        }
     }
 }
 
-/// Calculates the channel weight as (channel stake + 1) * (1 + R)
-/// where R is uniform random in (0,1]
+/// Assigns each channel a weight.
+/// The weight is randomized such that not always the same
+/// nodes get selected. This is necessary to achieve privacy.
+/// It also favors nodes with higher stake.
 pub struct RandomizedEdgeWeighting;
 
 impl EdgeWeighting<U256> for RandomizedEdgeWeighting {
-    fn calculate_weight(channel: &ChannelEntry) -> U256 {
-        const PATH_RANDOMNESS: f64 = 0.1;
-
-        let r = random_float() * PATH_RANDOMNESS;
-        let base = channel.balance.amount() + 1;
-
-        base.add(base.mul_f64(r).unwrap())
+    /// Multiply all channel stake with a random float in the interval (0,1].
+    /// Given that the floats are uniform, nodes with higher stake have a higher
+    /// probability of reaching a higher value.
+    ///
+    /// Sorting the list of weights thus moves nodes with higher stakes more
+    /// often to the front.
+    fn calculate_weight(channel: &ChannelEntry) -> Result<U256> {
+        let randomized_weight = channel
+            .balance
+            .amount()
+            .mul_f64(random_float())
```

Comment:
The `mul_f64` can only fail if the input arg is between not between `[0.0, 1.0]`. Since the input is randomly generated float in interval [0.0, 1), the `mul_f64` function is guaranteed not to fail here. So `unwrap/expect` (with a comment/message that this cannot fail) was appropriate.

Therefore please also remove the `ArithmeticError` enum variant.

Firstly, we should not allow random failures that we induce ourselves.
Secondly, the `mul_f64` function can be easily extended